### ISSUE TITLE
Simplified Linux support

### ISSUE THREAD

**Issue by superstator** [dlemstra/Magick.NET#150]
As a developer and end user, I'd like for Magick.NET on Linux to work out of the box just as it does on Windows. To fully meet that goal would require a Linux build that includes all the image and compression delegates that Windows does, and to statically link any support libraries not available out of the box on reference Linux distributions.

I have taken this as far as statically linking JPEG, PNG, TIFF, and WebP libraries, but things begin to get murky beyond that. OpenMP support is the biggest hangup, as libgomp1 is not a standard part of most Linux distros, including the microsoft/dotnet image. Statically linking it is possible, but not recommended, so that would still leave most people needing to install at least one system-level package before things "just work". I've gone through the motions of linking it in anyway, and it does work, but there exists the potential for weird conflicts or performance problems if used in combination with some other OpenMP-enabled software.

Thoughts?

**Comment by superstator** [comment]
Continuing to test against this, and the build I submitted in #149 has issues on Debian Stretch, as used in https://hub.docker.com/r/microsoft/dotnet/ - namely the easily installable packages for libjpeg and libwebp are not compatible. So perhaps that's an argument in favor of static linking that far down, at least.

**Comment by dlemstra** [comment]
It might be wiser to not enable OpenMP? Will take a look at your PR later this weekend.

**Comment by superstator** [comment]
I honestly wasn't sure where the dependency was being introduced and hadn't tried tracking it down yet. Is it just controlled by a preprocessor macro in Magick.NET.Native?

Don't worry to much about the PR yet, I will probably update it this week with more of a goldilocks build that includes image libraries, but not lower level stuff like zlib, etc. Working on a docker script right now to automatically verify ldd and dependencies on various platforms. Aiming for Ubuntu 16 - 18, Alpine, and CentOS 7.

**Comment by dlemstra** [comment]
Could you make sure you don't link anything that uses the GPL license? Would be nice to support as much delegates as possible. And it might be best to statically link even the libraries that come with the system?

**Comment by superstator** [comment]
I will keep that in mind. So far so good, at least with libjpeg-turbo, libtiff, libpng, and webp.

**Comment by superstator** [comment]
Just updated my fork and the PR. Now have it static linking almost everything. I added a Dockerfile to the ImageMagick folder which copies the .so's built by Tools\BuildCrossPlatform into a bunch of images and runs ldd against them. Everything is working out of the box on Ubuntu, Centos, and the Microsoft dotnet images.

Alpine is a challenge, as it uses musl libc. I tried just naively recompiling with musl-gcc, but no dice. I also tried the recommended glibc compatibility packages without any luck. glibc itself is GPL'd, so static linking it is not an option.

**Comment by hkoelewijn** [comment]
I am trying to get a dotnet test app to run in docker (Linux), but am not able to locate the dependencies:
        libMagickCore-7.Q16.so.5 => not found
        libMagickWand-7.Q16.so.5 => not found
        libjpeg.so.8 => not found
Sorry, I am not so well versed in linux, so I really don't know where to start to get these files. It looks like @superstator made some progress. Can you help me get under way?

**Comment by superstator** [comment]
Do you need any particular formats supported?

**Comment by hkoelewijn** [comment]
no, just the basics, jpeg, png, maybe bmp. I will need to do images comparison, scaling, cropping etc.

**Comment by superstator** [comment]
What we're doing for now is taking the binaries I built and manually creating our own NuGet package from them. I literally just unzip the official Magick.NET package, drop in my binaries, tweak the manifest, and stick it on our internal/corporate NuGet repository. So far so good, and we have a couple customers piloting an app now using containers built with that + the official .NET Core docker images. I'd be happy to send you a copy of our .nupkg if you want to give it a shot.

@dlemstra have you had a chance to look at that PR yet?

**Comment by hkoelewijn** [comment]
That would be great! You have a location where I can download that nuget package?

**Comment by superstator** [comment]
Shoot me an email, superstator@gmail.com and I'll see if I can get something to you. We are using the Q8 version, but it sounds like that should be fine for your purposes?

**Comment by hkoelewijn** [comment]
For anyone else wanting to run ImageMagick.NET in docker linux cintainers. Using the test nuget package with the native lib embedded worked. Hopefully this will be part of the official package soon!

**Comment by dlemstra** [comment]
@superstator I have been very busy with other things and haven't had time to take a proper look at your pull request.  But it looked promising when I took my first quick peek. I hope I will have time to take a look at it next weekend.

**Comment by superstator** [comment]
No worries, just wanted to check in. Let me know if there's anything I can do to help.

**Comment by dlemstra** [comment]
Left a review have fun :)

**Comment by dlemstra** [comment]
@superstator I just published a new version of Magick.NET that statically links ImageMagick. Would you mind helping me with linking fontconfig correctly? I needed to do some `sed` magick to disable some stuff to link ImageMagick. I don't understand why linking complains that some methods cannot be found.

@hkoelewijn Can you give the latest release a try?

**Comment by superstator** [comment]
Sure, I'll try to take a look at it this week.

**Comment by superstator** [comment]
@dlemstra How are you making line endings work on your end? I saw you removed all the dos2unix calls, but I'm finding I still need to run it to make things build reliably in Docker. Also, which freetype and fontconfig are you using? I didn't see either in amongst the ImageMagick repos.

**Comment by dlemstra** [comment]
I am building the Docker container on a Linux machine to avoid line ending issues. And I am cloning the fontconfig and freetype repos inside the build script.

**Comment by superstator** [comment]
Of course. Just sent you a PR.

**Comment by hkoelewijn** [comment]
I used the Q8-x64 version 7.4.2 with the Native dll from the 1st of march (from the Linux folder) and can confirm it works now.

**Comment by normj** [comment]
I was able to use the new version in an AWS Lambda function which is a Linux environment to resize an image. Nice work!
